osfoffline/utils/authentication.py

```diff
@@ -0,0 +1,201 @@
+import asyncio
+import datetime
+import json
+import logging
+
+import aiohttp
+import bcrypt
+import furl
+from PyQt5.QtWidgets import QMessageBox
+from sqlalchemy.orm.exc import NoResultFound, MultipleResultsFound
+from sqlalchemy.exc import SQLAlchemyError
+
+from osfoffline import settings
+from osfoffline.database_manager.db import session
+from osfoffline.database_manager.models import User
+from osfoffline.database_manager.utils import save
+from osfoffline.polling_osf_manager.remote_objects import RemoteUser
+
+
+def generate_hash(password, salt=None):
+    """ Generates a password hash using `bcrypt`.
+        Number of rounds for salt generation is 12.
+        Also used for checking passwords when salt
+        is passed in.
+
+    :param string password: password to hash
+    :param hash salt:       stored hash of a password for comparison.
+    :return:                hashed password
+    """
+    if not salt:
+        salt = bcrypt.gensalt(12)
+    return bcrypt.hashpw(
+        password.encode('utf-8'),
+        salt
+    )
+
+class AuthClient(object):
+    """Manages authorization flow """
+    def __init__(self):
+        self.failed_login = False
+
+    @asyncio.coroutine
+    def _authenticate(self, username, password):
+        """ Tries to use standard auth to authenticate and create a personal access token through APIv2
+
+            :return: personal_access_token or None
+        """
+        token_url = furl.furl(settings.API_BASE)
+        token_url.path.add('/v2/tokens/')
+        token_request_body = '{"data": {"attributes": {"name": "OSF-Offline ' + \
+            str(datetime.date.today()) + \
+            '", "scopes": "' + settings.APPLICATION_SCOPES + \
+            '"}, "type": "tokens"}}'
```

Comment:
Avoid building JSON via string concatenation- there are a variety of ways that could break. Instead, build a python dictionary and then pass `json.dumps(token_request_body)` in as the data argument of `aiohttp.request`